midend/flattenInterfaceStructs.h

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2018 VMware, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _MIDEND_FLATTENINTERFACESTRUCT_H_
+#define _MIDEND_FLATTENINTERFACESTRUCT_H_
+
+#include "ir/ir.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+
+namespace P4 {
+
+/**
+   Describes how a type is replaced: the new type to
+   replace it and how each field is renamed.
+*/
+struct Replacement : public IHasDbPrint {
+    // The keys are strings of the form field.field.field.  If
+    // something is mapped to the null string it does not really have
+    // a replacement: this is true for fields whose types are structs.
+    std::map<cstring, cstring> fieldNameRemap;
+    const IR::Type* type;
+    virtual void dbprint(std::ostream& out) const {
+        out << type;
+    }
+};
+
+/**
+ * Maintains a map between an original struct type and its
+ * replacement.
+ */
+struct NestedStructMap {
+    P4::ReferenceMap* refMap;
+    P4::TypeMap* typeMap;
+
+    ordered_map<const IR::Type*, Replacement*> replacement;
+
+    NestedStructMap(P4::ReferenceMap* refMap, P4::TypeMap* typeMap):
+            refMap(refMap), typeMap(typeMap)
+    { CHECK_NULL(refMap); CHECK_NULL(typeMap); }
+    void createReplacement(const IR::Type_Struct* type);
+    Replacement* getReplacement(const IR::Type* type) const
+    { return ::get(replacement, type); }
+    bool empty() const { return replacement.empty(); }
+};
+
+/**
+Find the types to replace and insert them in the nested struct map.
+ */
+class FindTypesToReplace : public Inspector {
+    NestedStructMap* map;
+ public:
+    explicit FindTypesToReplace(NestedStructMap* map): map(map) {
+        setName("FindTypesToReplace");
+        CHECK_NULL(map);
+    }
+    bool preorder(const IR::Declaration_Instance* inst) override;
+};
+
+/**
+This pass transforms the type signatures of instantiated controls,
+parsers, and packages.  It does not transform methods, functions or
+actions.  It starts from package instantiations: every type argument
+that is a nested structure is replaced with "simpler" flatter type.
+
+Should be run after the NestedStructs pass.
+
+struct S { bit b; }
+struct T { S s; }
+
+control c(inout T arg) {
+    apply {
+        ... arg.s.b ...
+    }
+}
+
+control proto<V>(inout V arg);
+package top<V>(proto<V> ctrl);
+
+top<T>(c()) main;
+
+This is transformed into:
+
+struct S { bit b; }
+struct T { S s; }
+struct Tflat {
```

Comment:
What is the reason for not replacing the nested structs in place? By generating a new type and replacing it in the program, you affect the control plane code generation (when done past midend which some backends may do) and make debugging a lot harder.

Reply:
You are really asking why I am not keeping the same type name. This is really a new type, it is not the same as the old type, so I liked the idea of having a new name. I don't think there isn't any reason we can't keep the old name. I will try the patch that way.

I am not aware of any place where the control-plane API depends on type names. I.e., I don't know if you alpha-rename types in a P4 program whether the control-plane API could change at all. But this is indeed not part of the P4 contract, so it is conceivable that some APIs could depend on type names.

The fact that the type structure changes is actually a much more disruptive change than the renaming of the type. I hope that is not going to break anything. That's one reason it took me so long to implement this, and I am still not sure that this is 100% correct (i.e., there are no source programs where this transformation could generate an output program that is not equivalent).

Reply:
Actually, the transformation is not enough, some of the failing test cases show instances where it fails. So more work is needed.

Reply:
> I am not aware of any place where the control-plane API depends on type names. I.e., I don't know if you alpha-rename types in a P4 program whether the control-plane API could change at all. But this is indeed not part of the P4 contract, so it is conceivable that some APIs could depend on type names.

Part of the work done on the P4Runtime API to allow arbitrary types to be used as Digest extern message parameters, and in the future other places as well (e.g. hopefully action parameters, table keys, Register array element contents), uses type names in the control plane API to name those types.  I do not know the details of these changes and which pass they might appear in the compiler to know whether that could affect the output files used by the P4Runtime API, but wanted to address your comment, so you were aware of this potential issue.

Reply:
The PR is developing a new midend pass.  The P4Runtime API are generated by a backend once the backend code has completed running the frontend module.  See frontend.run(..,..) in p4c/backends/bmv2/simple_switch/main.cpp.  Thus, the new flatten midend pass of this PR (or any other midend pass) is run after the P4Runtime API are generated.

Reply:
Yes, but there are compilers out there that do the API generation after mid-end.

Reply:
The ebpf back-end generates the control-plane API in the back-end. It does not use P4Runtime.
It does not need this pass, though.
@cc10512 may have other examples, he is the one who said this.
I am not working on any proprietary compilers for P4, all the code I write is on github, so there is nothing else that I know about this subject.

Reply:
> Yes, but there are compilers out there that do the API generation after mid-end.

Hmm, Antonin sent out an email to p4-dev in a discussion with me saying "it was a careful decision where to insert control-plane api generation."   See

http://lists.p4.org/pipermail/p4-dev_lists.p4.org/2018-September/003636.html

It would be good to understand why compilers are diverging.  I see even the p4test backend generates P4Runtime API after frontend.run() and before midend.process() is called.

Reply:
Compilers are not diverging, they support different architectures with different constraints. When you run eBPF locally in the Linux kernel often it does not make sense to have a remote control-plane that communicates using RPC. And you need a local control-plane anyway. 

The P4 spec does not mandate a specific compiler architecture; it only requires some entities to be exposed to the control-plane, i.e., tables and actions. As long as these are done properly it should not matter where or how the control-plane API is being generated. P4Runtime is *a* control-plane API, but not the only one that can work with P4 programs. That's exactly why P4Runtime is a separate spec. 

In our implementation we are indeed careful when we add passes to the front-end: these passes should not cause any changes in the control-plane API that will be generated. No such contract is promised by mid-end passes, but if you construct your mid-end carefully you can preserve this property after the mid-end as well.

Reply:
> Compilers are not diverging, they support different architectures with different constraints. When you run eBPF locally in the Linux kernel often it does not make sense to have a remote control-plane that communicates using RPC. And you need a local control-plane anyway.
> 
> The P4 spec does not mandate a specific compiler architecture; it only requires some entities to be exposed to the control-plane, i.e., tables and actions. As long as these are done properly it should not matter where or how the control-plane API is being generated. P4Runtime is _a_ control-plane API, but not the only one that can work with P4 programs. That's exactly why P4Runtime is a separate spec.
> 
> In our implementation we are indeed careful when we add passes to the front-end: these passes should not cause any changes in the control-plane API that will be generated. No such contract is promised by mid-end passes, but if you construct your mid-end carefully you can preserve this property after the mid-end as well.

I was only going why what Antonin said that it's a hard rule where to insert P4Runtime generation.    However, I understand your perspective as well.